### DISCUSSION TITLE
중간발표 준비

### DISCUSSION THREAD

**Discussion by parkjihwanjay** [parkjihwanjay/naver-course-project-12#9] [General]
# 들어갈 내용

## 요구사항
- 해당 프로젝트에 **필수 기술** 요소들이 어떻게 들어가는지 소개 해야함

### 필수 기술 요소들
- 이벤트 처리
  - [React 합성 이벤트](https://ko.reactjs.org/docs/events.html) 사용
  - 이벤트 핸들러를 Container에서 내려주는 패턴
  - 300ms 내에 한번 이상 발생하는 이벤트의 경우 상황에 따라 Event Debounce 또는 Event Throttling 사용
    - DragDrop API를 사용하기 때문에 사용할 일은 없을듯..?
- 상태관리
  - Redux, Redux toolkit 사용
  - Card 정보를 Redux로 관리
  - Container, Presenter 패턴을 활용하여 dispatch하는 함수를 Presenter에게 props로 내려주는 패턴 사용
  - state, dispatch를 한번에 받도록 custom hook 사용

- API 서버 연동
  - Rest API 사용
  - 자체 서버 구성(Express)
  - api 디렉토리에 api 요청들을 정리
    - try, catch를 한번 wrapping해서 [result, error]를 반환하는 함수 작성
  - 서버 주소, 포트 번호는 환경 변수로 저장
  - StatusCode는 상수로 저장
  - Request, Response를 위한 DTO 타입 정의

## 추가 사항
- Stack
  - 스택을 사용하는 간략한 이유
- Container, Presenter 패턴
  - 왜 쓰는지, 프로젝트에 어떤 점이 이 패턴에 맞는지
    - 문제가 생겼을 때 어디를 봐야하는지 알기 쉽다
      - ex) UI 문제 -> Presenter, 비지니스 로직 문제 -> Container
    - 비지니스 로직을 Container에서 한번에 볼 수 있다.
- 작업 순서
  - 프론트엔드 완성 후 백엔드 작업
  - 프로젝트 특성상 프론트엔드가 중요하기 때문에
- Branch, Commit 규칙
  - [Wiki Development Rule](https://github.com/parkjihwanjay/naver-course-project-12/wiki/Development-Rule) 참고
- Github Flow, Git flow
- Drag Drop 어떻게 구현할지

# 중요한 점
현재 프로젝트 주제와 상황에 맞게 이유를 설명할 것

**Comment by parkjihwanjay** [comment]
## Question
by 정성원

1. DragDrop API랑 event debounce/throttling이 어떻게 연관있는지 궁금합니다
2. api를 Express로 구성한 다음 nextjs의 api 디렉토리에서 어떻게 하는 것인지 잘 모르겠습니다..!
3. StatusCode는 상수로 저장?
4. Request, Response를 위한 DTO 타입 정의
5. Git flow 사용? (master, dev, feature, setting, ...)

**Reply by parkjihwanjay** [reply]
1. DragDrop API 자체가 mouse event를 내부적으로 사용하는거 같아서 mouse event 자체에 debounce나 throttling을 적용하기는 힘들 것 같은데 혹시 성능 이슈가 발생하면 mouse event를 자체적으로 사용하고 debounce나 throttling을 적용해야하지 않나?하는 의미로 적었습니다!

2. 저도 좀 더 공부해봐야 알거 같은데 express로 어플리케이션 서버를 구축하면 next.js에서 제공하는 api 서버는 사용하지 않아도 될 것 같습니다. 제가 말한 api 디렉토리는 next.js에서 제공해주는 pages/api가 아니라 프로젝트 루트에 저희가 만드는 api 디렉토리를 의미합니다..!

3. StatusCode와 해당 code에 맞는 message를 상수로 저장하면 좋은데 어느 디렉토리에 어떤 파일로 저장할지는 논의해보면 좋을 것 같습니다!

4. DTO라는 용어를  프론트에서 api 요청과 응답으로 사용해도 되는지 모르겠지만 Request와 Response 타입을 미리 정의해두자는 의미로 적었습니다..! 개인적으로는 interfaces라는 디렉토리를 따로 만들어서 정의해두면 좋겠다는 생각입니다!

5. dev 브랜치를 기본으로 두고 기능 추가를 할 경우 feature/#이슈번호/간략한 설명의 형식으로 브랜치를 만들어서 작업 진행 후 push 한다 이 정도로만 적어도 좋을 것 같습니다! (이후 과정은 Github Flow라고 판단되네요)

**Reply by SeongwonChung** [reply]
답변 감사합니다!

**Comment by parkjihwanjay** [comment]
추가로 
1. redux-thunk를 활용하여 비동기 로직 처리
2. redux-toolkit을 활용한 ducks 패턴 활용

들어가면 좋을 것 같습니다..!
